platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbCompactionCoordinator.java

```diff
@@ -54,31 +55,34 @@ class MerkleDbCompactionCoordinator {
 
     private static final Logger logger = LogManager.getLogger(MerkleDbCompactionCoordinator.class);
 
-    /**
-     * Since {@code com.swirlds.platform.Browser} populates settings, and it is loaded before any
-     * application classes that might instantiate a data source, the {@link ConfigurationHolder}
-     * holder will have been configured by the time this static initializer runs.
-     */
-    private static final MerkleDbConfig config = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
     /**
      * An executor service to run compaction tasks.
      */
-    private static final ExecutorService compactionExecutor;
+    private static final AtomicReference<ExecutorService> compactionExecutor = new AtomicReference<>();
 
-    static {
-        compactionExecutor = new ThreadPoolExecutor(
-                config.compactionThreads(),
-                config.compactionThreads(),
-                50L,
-                TimeUnit.MILLISECONDS,
-                new LinkedBlockingQueue<>(),
-                new ThreadConfiguration(getStaticThreadManager())
-                        .setThreadGroup(new ThreadGroup("Compaction"))
-                        .setComponent(MERKLEDB_COMPONENT)
-                        .setThreadName("Compacting")
-                        .setExceptionHandler(
-                                (t, ex) -> logger.error(EXCEPTION.getMarker(), "Uncaught exception during merging", ex))
-                        .buildFactory());
+    static ExecutorService getCompactionExecutor() {
+        ExecutorService exec = compactionExecutor.get();
+        if (exec == null) {
```

Comment:
Minor: I hope there's not a thread race, today or in the future. I'd put `synchronized` on the method to be safe. While I see that you're calling compareAndSet() below, but then you have to shut down the newly created one. This seems like an extra complexity that could be avoided altogether. If the `synchronized` is being avoided for speed, then perhaps using the holder pattern here would be a better and cleaner choice (which in its simplest form would move the `synchronized{}` inside this method and use a double-check pattern.)

Reply:
You're right, it isn't synchronized because of `compareAndSet()`, and it's done for performance reasons. It's better to deal with very unlikely event of shutting down a redundant executor than to pay for synchronization on every call.

On the other hand... This method is called just a couple times for every flush, multiplied by the number of virtual maps. In other words, not performance critical at all. The chances are very low that the method is called in parallel from multiple threads, too.

@OlegMazurov What do you think? Should it be synchronized (and the field can be changed from an atomic reference to just a reference), or the current approach with CAS is better?

Reply:
`volatile compactionExecutor` and double checking is the usual pattern here.